Use alpine's JRE instead of AdoptOpenJDK

username-is-too-long opened the pull request:
[AdoptOpenJDK](https://github.com/AdoptOpenJDK/openjdk-docker) now has a DEPRECATION NOTICE warning, and Alpine has its own packaged openjdk/jre that uses musl libc, so lines 23-35 and 55-86 could be replaced.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

MarkEWaite commented:
I recommend against this change based on the comments in jenkinsci/docker#1185 and in jenkinsci/docker#1186.  I'm open to reconsider based on guidance from @timja and @dduportal , but the transition to Eclipse Temurin (AdoptOpenJDK) has been a positive change.  One Java provider for all our Docker images with synchronized updates across all our platforms (AMD64, Arm64, PowerPC, and s390x).

timja commented:
I’ll look at this closer later but I think it might be better to take this as a temporary switch until temurin provides a musl based build,

see https://github.com/adoptium/containers/issues/1 for background

username-is-too-long commented:
> see [adoptium/containers#1](https://github.com/adoptium/containers/issues/1) for background

Oh nice, I didn't see they were moving to musl.

timja commented:
> > see [adoptium/containers#1](https://github.com/adoptium/containers/issues/1) for background
> 
> Oh nice, I didn't see they were moving to musl on Alpine.

It looks like awhile away, they first need to TCK 16+ and then someone needs to backport to 11 and 8. They also need to setup some infrastructure for it.


I wouldn’t expect it to be quick but I think we can swap to this in the meantime

timja commented:
@username-is-too-long the locale appears to not be set to a utf-8 one can you check it?

there was some code in the glibc section doing that, it might just be a matter of switching the locale line to:

```
ENV LANG C.UTF-8
```

but I don't think that worked last time I tried, see:
https://grrr.tech/posts/2020/add-locales-to-alpine-linux-docker-image/

username-is-too-long commented:
> @username-is-too-long the locale appears to not be set to a utf-8 one can you check it?
> 
> there was some code in the glibc section doing that, it might just be a matter of switching the locale line to:
> 

`/usr/glibc-compat/bin/locale` doesn't exist without without glibc, I added musl-locales to provide `locale`, and removed that alpine case in tests.bats.

timja commented on 11/alpine/Dockerfile:
```diff
@@ -49,54 +36,14 @@ ARG AGENT_WORKDIR=/home/${user}/agent
-    && rm -rf /tmp/*.apk /tmp/gcc /tmp/gcc-libs.tar* /tmp/libz /tmp/libz.tar.xz /var/cache/apk/*
-
-RUN apk add --update --no-cache curl bash git git-lfs openssh-client openssl procps \
+RUN apk add --update --no-cache curl bash git git-lfs musl-locales openjdk11-jre-headless openssh-client openssl procps \
```
we should still be running `jlink` to get a trimmed down runtime, it should be smaller than the `jre`

username-is-too-long replied:
You're right, saved about 80mb

timja commented:
(I sorted the build separately in a different PR, green now 😄 )

timja commented on 11/alpine/Dockerfile:
```diff
@@ -20,12 +20,15 @@
 # Generate smaller java runtime without unneeded files
 # for now we include the full module path to maintain compatibility
-# while still saving space (approx 200mb from the full distribution)
+# while still saving space (approx 200mb from the full distribution)                                 
```
```suggestion
# while still saving space (approx 200mb from the full distribution)
```

timja commented on 11/alpine/Dockerfile:
```diff
@@ -20,12 +20,15 @@
-# while still saving space (approx 200mb from the full distribution)
+# while still saving space (approx 200mb from the full distribution)                                 
 RUN jlink \
+	 --module-path /usr/lib/jvm/default-jvm/jmods \
```
```suggestion
             --module-path /usr/lib/jvm/default-jvm/jmods \
```

timja commented on 11/alpine/Dockerfile:
```diff
@@ -49,51 +52,15 @@ ARG AGENT_WORKDIR=/home/${user}/agent
   && rm -rf /tmp/*.apk /tmp/gcc /tmp/gcc-libs.tar* /tmp/libz /tmp/libz.tar.xz /var/cache/apk/*
 
-ENV JAVA_HOME=/opt/java/openjdk
+ENV JAVA_HOME=/opt/java/openjdk        
```
```suggestion
ENV JAVA_HOME=/opt/java/openjdk        
```

timja commented on 8/alpine/Dockerfile:
```diff
@@ -20,10 +20,7 @@
-# There is no official Alpine images at the moment.
-# Needs upgrade when/if there is an official alpine image.
-FROM adoptopenjdk/openjdk8:jdk8u302-b08-alpine
+FROM alpine:3.14.1
```
```suggestion
FROM alpine:3.14.2
```

timja commented on 11/alpine/Dockerfile:
```diff
@@ -20,12 +20,15 @@
 #  THE SOFTWARE.
 
-FROM adoptopenjdk/openjdk11:jdk-11.0.12_7-alpine AS jre-build
+FROM alpine:3.14.1 as jre-build
```
```suggestion
FROM alpine:3.14.2 as jre-build
```

timja commented on 11/alpine/Dockerfile:
```diff
@@ -20,12 +20,15 @@
 # for now we include the full module path to maintain compatibility
 # while still saving space (approx 200mb from the full distribution)
 RUN jlink \
+             --module-path /usr/lib/jvm/default-jvm/jmods \
```
```suggestion
         --module-path /usr/lib/jvm/default-jvm/jmods \
```
